[filebeat][Azure Blob Storage] - Improved documentation (#41252)
elastic/beats: x-pack/filebeat/docs/inputs/input-azure-blob-storage.asciidoc

    describing said error.

[id="supported-types"]

> NOTE: NOTE: Currently only `JSON`, `NDJSON` and `CSV` are supported blob/file formats. Blobs/files may be also be gzip compressed.

Reviewer: Do we need  "NOTE: NOTE:"‽

```suggestion
`JSON`, `NDJSON` and `CSV` are supported blob/file formats. Blobs/files may be also be gzip compressed.
```

Author: Wanted to highlight this since users could potentially miss this if not highlighted via a NOTE.

Reviewer: Still one to go. In this context, I don't believe that the NOTE: will make any difference to whether a reader will pay attention to the information. If you want to ensure it's noticed, I'd suggest giving it a heading. The use of "NOTE" marks something as incidental but something to consider. If it's more important than that, which it appears to be according to what you've said, then it needs a section.

    describing said error.

[id="supported-types"]
NOTE: `JSON`, `NDJSON` and `CSV` are supported blob/file formats. Blobs/files may be also be gzip compressed.

> As for authentication types, we currently have support for `shared access keys`, `connection strings` and `Microsoft Entra ID RBAC`.

Reviewer: "As for" is conversational English. Suggest
```suggestion
`shared access keys`, `connection strings` and `Microsoft Entra ID RBAC` authentication types are supported.
```

Reviewer: s/NOTE://

[float]
==== `max_workers`

> This attribute defines the maximum number of workers (goroutines / lightweight threads) are allocated in the worker pool (thread pool) for processing jobs which read the contents of files. This attribute can be specified both at the root level of the configuration and at the container level. Container level values override the root level values if both are specified. Larger number of workers do not necessarily improve of throughput, and this should be carefully tuned based on the number of files, the size of the files being processed and resources available. Increasing `max_workers` to very high values may cause resource utilization problems and can lead to a bottleneck in processing. Usually a maximum cap of `2000` workers is recommended. A very low `max_worker` count will drastically increase the number of network calls required to fetch the blobs, which can cause a bottleneck in processing.

Reviewer: ```suggestion
This attribute defines the maximum number of workers allocated to the worker pool for processing jobs which read file contents. It can be specified both at the root level of the configuration, and at the container level. Container level values override root level values if both are specified. Larger number of workers do not necessarily improve throughput, and this should be carefully tuned based on the number of files, the size of the files being processed and resources available. Increasing `max_workers` to very high values may cause resource utilization problems and may lead to bottlenecks in processing. Usually a maximum of `2000` workers is recommended. A very low `max_worker` count will drastically increase the number of network calls required to fetch the blobs, which may cause a bottleneck in processing.
```

> NOTE: The value of `max_workers` is tied to the `batch_size` currently to ensure even distribution of workloads across all goroutines. This ensures that the input is able to process the files in an efficient manner. This `batch_size` determines how many blobs will be fetched in one single call. The `max_workers` value should be set based on the number of files to be read, the resources available and the network speed. For example,`max_workers=3` would mean that every pagination request a total number of `3` blobs are fetched and distributed among `3 goroutines`, `max_workers=100` would mean `100` blobs are fetched in every pagination request and distributed among `100 goroutines`.

Reviewer: ```suggestion
The value of `max_workers` is tied to the `batch_size` currently to ensure even distribution of workloads across all workers. This ensures that the input is able to process the files in an efficient manner. This `batch_size` determines how many blobs will be fetched in one single call. The `max_workers` value should be set based on the number of files to be read, the resources available and the network speed. For example,`max_workers=3` would mean that every pagination request a total number of `3` blobs are fetched and distributed among `3 workers`, `max_workers=100` would mean `100` blobs are fetched in every pagination request and distributed among `100 workers`.
```

I don't understand what "tied' means here. I don't think that it means what it is intended to mean though. "tied" should mean that we do work to ensure that it is modified based on the value of `batch_size` which AFAIUI we do not.

I've replaced all instances of goroutines with workers since a goroutine is a technical detail that a user should not be interested in. They want to know (at most, likely) that there is concurrency and worker embodies that.

Author: I feel like we might change the implementation on batch size later down the road since there can be scenarios when having a 1:1 distribution might not be ideal. Also decoupling this would give the users more flexibility in controlling their desired batch size without messing with the worker count. What do you think @efd6 ?

Reviewer: I don't yet know what the terms are here, so I can't answer that.

Author: updated this section

    - regex: '/Security-Logs/'
----

> The `file_selectors` operation is performed within the agent locally, hence using this option will cause the agent to download all the files and then filter them. This can cause a bottleneck in processing if the number of files are very high. It is recommended to use this attribute only when the number of files are limited or ample resources are available.

Reviewer: ```suggestion
The `file_selectors` operation is performed within the agent locally. The agent will download all the files and then filter them based on the `file_selectors`. This can cause a bottleneck in processing if the number of files are very high. It is recommended to use this attribute only when the number of files are limited or ample resources are available.
```

    timestamp_epoch: 1627233600
----

> The Azure Blob Storage APIs don't provide a direct way to filter files based on the timestamp, so the input will download all the files and then filter them based on the timestamp. This can cause a bottleneck in processing if the number of files are very high. It is recommended to use this attribute only when the number of files are limited or ample resources are available.

Reviewer: ```suggestion
The Azure Blob Storage APIs don't provide a direct way to filter files based on timestamp, so the input will download all the files and then filter them based on the timestamp. This can cause a bottleneck in processing if the number of files are very high. It is recommended to use this attribute only when the number of files are limited or ample resources are available.
```
